packages/affiliates/libs/affiliates/dappmining_v2.js

```diff
@@ -0,0 +1,278 @@
+      return (
+        stream
+          .reduce({ lastBlock: startBlock, balancesSum }, ({ lastBlock, balancesSum }, event) => {
+            if (event.blockNumber > lastBlock) {
+              // maintains the weighed sum of all balances for every block
+              balancesSum = sumBalances(balances.tokens, event.blockNumber - lastBlock, balancesSum);
+              lastBlock = event.blockNumber;
+            }
+            // maintains the latest state of all balances for all users
+            balances.handleEvent(event);
+            return { lastBlock, balancesSum };
+          })
+          // finalizes the balances sum by weighting any unprocessed blocks to the end of the timeframe
+          .map(({ lastBlock, balancesSum }) => {
+            return sumBalances(balances.tokens, endBlock - lastBlock, balancesSum);
+          })
+      );
+    };
+  }
+
+  // Helper function to process rewards. Use this for testing business logic with custom data.
+  async function processRewardData({ attributions, balances, whitelist, totalRewards }) {
+    assert(totalRewards, "requires total rewards");
+    const percentages = calculateBlockReward({
+      attributions,
+      balances,
+      whitelist
+    });
+
+    return Object.entries(percentages).reduce((result, [key, value]) => {
+      if (value == 0) return result;
+      result[key] = calculatePercent(totalRewards, value);
+      return result;
+    }, {});
+  }
+
+  // Main entrypoint to kick off reward calculations. Returns an object with rewards keyed by address and value.
+  async function getRewards({ empAddress, defaultAddress, startTime, endTime, whitelist, totalRewards, firstEmpDate }) {
+    assert(empAddress, "requires emp address for reward calculation");
+    assert(defaultAddress, "requires defaultAddress for untagged rewards");
+    assert(startTime >= 0, "requires startTime >= 0");
+    assert(endTime > 0 && endTime > startTime, "requires endTime >= 0 && > startTime");
+    assert(whitelist && whitelist.length, "requires an array of whitelisted payout addresses");
+    assert(totalRewards > 0, "requires totalRewards above 0 to be shared across dapps");
+
+    // use firstEmpDate as a history cutoff when querying for events. We can safely say no emps were deployed before Jan of 2020.
+    firstEmpDate = firstEmpDate || moment("2020-01-01", "YYYY-MM-DD").valueOf();
+
+    // Pull external data
+    const attributionStream = getAttributionStream(empAddress, firstEmpDate, endTime);
+    const eventStream = getEventStream(empAddress, firstEmpDate, endTime);
+    const startBlock = await getStartBlock(startTime);
+    const endBlock = await getEndBlock(endTime);
+
+    // Wire up streaming logic
+    const attributionPromise = attributionStream
+      .through(
+        ProcessAttributionStream({
+          startBlock: startBlock.number,
+          endBlock: endBlock.number,
+          attributions: EmpAttributions(empAbi, defaultAddress, AttributionLookback())
```

Comment:
EmpAttribution process now optionally takes the attribution model, so we pass it the new lookback model to be populated by tagged mint events.